docs/tutorial/codecs.txt

```diff
@@ -0,0 +1,433 @@
+           private readonly DateTimeCodec $dateTimeCodec = new DateTimeCodec(),
+       ) {}
+
+       // Other code omitted for brevity
+       public function decode($value): Person
+       {
+           if (! $this->canDecode($value)) {
+               throw UnsupportedValueException::invalidDecodableValue($value);
+           }
+
+           $person = new Person($value->get('name'));
+           $person->id = $value->get('_id');
+           $person->createdAt = $this->dateTimeCodec->decode($value->get('createdAt'));
+
+           return $person;
+       }
+
+       public function encode($value): MongoDB\BSON\Document
+       {
+           if (! $this->canEncode($value)) {
+               throw UnsupportedValueException::invalidEncodableValue($value);
+           }
+
+           return MongoDB\BSON\Document::fromPHP([
+               '_id' => $value->id,
+               'name' => $value->name,
+               'createdAt' => $this->dateTimeCodec->encode($value->createdAt),
+           ]);
+       }
+   }
+
+Handling embedded documents
+---------------------------
+
+The previous example showed how to handle a single document. However, sometimes you want to handle fields that contain
+embedded documents. To show this, let's create an address document that we'll be embedding into our ``Person`` document.
+To ensure consistency, we're going to make this a readonly class:
+
+.. code-block:: php
+
+   <?php
+
+   final readonly class Address
+   {
+       public function __construct(
+           public string $street,
+           public string $postCode,
+           public string $city,
+           public string $country,
+       ) {
+   }
+
+We can now create a document codec for this class:
+
+.. code-block:: php
+
+   <?php
+
+   final class AddressCodec implements MongoDB\Codec\DocumentCodec
+   {
+       // Other code omitted for brevity
+       public function decode($value): Person
+       {
+           if (! $this->canDecode($value)) {
+               throw UnsupportedValueException::invalidDecodableValue($value);
+           }
+
+           return new Address(
+               $value->get('street'),
+               $value->get('postCode'),
+               $value->get('city'),
+               $value->get('country'),
+           );
+       }
+
+       public function encode($value): MongoDB\BSON\Document
+       {
+           if (! $this->canEncode($value)) {
+               throw UnsupportedValueException::invalidEncodableValue($value);
+           }
+
+           return MongoDB\BSON\Document::fromPHP([
+               'street' => $value->street,
+               'postCode' => $value->postCode,
+               'city' => $value->city,
+               'country' => $value->country,
+           ]);
+       }
+   }
+
+This codec is quite similar to the ``PersonCodec`` we had before, except that we're not adding an identifier to it. The
+creation of the object also differs, as we have to pass all the fields to the constructor.
```

Comment:
```suggestion
creation of the object also differs, as we have to pass all fields to the constructor.
```

Either "all fields" or "all of the fields", but not "all the fields".

Is this only necessary because the class is read-only? Did you want to elaborate on what you meant by "ensuring consistency" with this design?

Reply:
Removed the entire paragraph: the `PersonCodec` was changed to also pass all fields in the constructor, and the first sentence feels lost by itself.